src/bugpoint.rs

```diff
@@ -0,0 +1,312 @@
+                            .template("{bar:60} {prefix:40} {pos:>4}/{len:>4} {msg}"),
+                    );
+                    progress.set_prefix(&format!("pass {} phase {}", pass_idx, msg));
+                    continue 'inner_loop;
+                }
+                PhaseStepResult::Finished => {
+                    progress.set_message("done");
+                    progress.finish();
+                    break 'inner_loop;
+                }
+            };
+
+            progress.set_message(&msg);
+
+            match check_for_crash(isa, &func2) {
+                Res::Succeed => {
+                    // Shrinking didn't hit the problem anymore, discard changes.
+                    //progress.println("succeeded");
+                    continue;
+                }
+                Res::Verifier(err) => {
+                    // Shrinking produced invalid clif, discard changes.
+                    //progress.println(format!("verifier error {}", err));
+                    continue;
+                }
+                Res::Panic => {
+                    // Panic remained while shrinking, make changes definitive.
+                    func = func2;
+                    if shrinked {
+                        was_reduced = true;
+                        if verbose {
+                            progress.println(format!("{}: shrink", msg));
+                        }
+                    } else {
+                        if verbose {
+                            progress.println(format!("{}: replace", msg));
+                        }
+                    }
+                }
+            }
+        }
+
+        if !was_reduced {
+            // No new shrinking opportunities have been found this pass. This means none will ever
+            // be found. Skip the rest of the passes over the function.
+            break 'outer_loop;
+        }
+    }
+
+    println!("\n{}", func);
+
+    println!(
+        "{} ebbs {} insts -> {} ebbs {} insts",
+        orig_ebb_count,
+        orig_inst_count,
+        ebb_count(&func),
+        inst_count(&func)
+    );
+}
+
+enum Res {
+    Succeed,
+    Verifier(String),
+    Panic,
+}
+
+fn check_for_crash(isa: &TargetIsa, func: &Function) -> Res {
+    let mut context = Context::new();
+    context.func = func.clone();
+
+    let mut relocs = PrintRelocs::new(false);
+    let mut traps = PrintTraps::new(false);
+    let mut mem = vec![];
+
+    use std::io::Write;
+    std::io::stdout().flush().unwrap(); // Flush stdout to sync with panic messages on stderr
+
+    match std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
+        if let Err(err) = cranelift_codegen::verifier::verify_function(&func, isa) {
+            Some(err)
+        } else {
+            None
+        }
+    })) {
+        Ok(Some(err)) => return Res::Verifier(err.to_string()),
+        Ok(None) => {}
+        Err(err) => {
+            // FIXME prevent verifier panic on removing ebb1
```

Comment:
FIXME here.

Reply:
The panic is: "EBB node must have one reachable predecessor".

If I run `cargo run compile` on the ir shown when inserting `println!("{}", func);` here, it gives a normal verifier error:

```
[...]
    jump ebb1
;   ^~~~~~~~~
; error: inst102: invalid ebb reference ebb1


ebb2:
    trap user0
[...]
```

Reply:
I think there's a method to recompute successors and predecessors after an EBB has been introduced or removed, so using it might avoid this verifier issue.

Reply:
I believe you are refering to [`Context::compute_cfg`](https://docs.rs/cranelift-codegen/0.30.0/cranelift_codegen/struct.Context.html#method.compute_cfg). Calling it wont solve it, as the `Context` is not reused.

Reply:
nit: One way would be to only reduce on panics which are keeping a similar error message. This way, even if the panic message changes, it would not be accepted as a reducing case.

Reply:
I prefer to just fix the verifier. Having other panics being allowed would make it easier to find panics which you didnt even know existed. Maybe even make bugpoint reduce all different panics it found.